providers/implementations/keymgmt/mlkem_kmgmt.c

```diff
@@ -0,0 +1,585 @@
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/core_dispatch.h>
+#include <openssl/core_names.h>
+#include <openssl/params.h>
+#include <openssl/err.h>
+#include <openssl/proverr.h>
+#include <openssl/rand.h>
+#include <openssl/self_test.h>
+#include "internal/param_build_set.h"
+#include <openssl/param_build.h>
+#include "prov/mlkem.h"
+#include "prov/implementations.h"
+#include "prov/providercommon.h"
+#include "prov/provider_ctx.h"
+#include "prov/securitycheck.h"
+#include <assert.h>
+
+#define BUFSIZE 1000
+#if defined(NDEBUG) || defined(OPENSSL_NO_STDIO)
+static void debug_print(char *fmt, ...)
+{
+}
+#else
+static void debug_print(char *fmt, ...)
+{
+    char out[BUFSIZE];
+    va_list argptr;
+
+    va_start(argptr, fmt);
+    vsnprintf(out, BUFSIZE, fmt, argptr);
+    va_end(argptr);
+    if (getenv("TEMPLATEKM"))
+        fprintf(stderr, "TEMPLATE_KM: %s", out);
+}
+#endif
+
+static void print_hex(const uint8_t *data, int len, const char *msg)
+{
+#ifndef NDEBUG
+    if (msg)
+        printf("%s: \n", msg);
+    BIO_dump_fp(stdout, data, len);
+    printf("\n\n");
+#endif
+}
+
+static OSSL_FUNC_keymgmt_new_fn mlkem_new;
+static OSSL_FUNC_keymgmt_free_fn mlkem_free;
+static OSSL_FUNC_keymgmt_gen_init_fn mlkem_gen_init;
+static OSSL_FUNC_keymgmt_gen_fn mlkem_gen;
+static OSSL_FUNC_keymgmt_gen_cleanup_fn mlkem_gen_cleanup;
+static OSSL_FUNC_keymgmt_gen_set_params_fn mlkem_gen_set_params;
+static OSSL_FUNC_keymgmt_gen_settable_params_fn mlkem_gen_settable_params;
+static OSSL_FUNC_keymgmt_get_params_fn mlkem_get_params;
+static OSSL_FUNC_keymgmt_gettable_params_fn mlkem_gettable_params;
+static OSSL_FUNC_keymgmt_set_params_fn mlkem_set_params;
+static OSSL_FUNC_keymgmt_settable_params_fn mlkem_settable_params;
+static OSSL_FUNC_keymgmt_has_fn mlkem_has;
+static OSSL_FUNC_keymgmt_match_fn mlkem_match;
+
+/* implement only when encode/decode logic becomes required/standardized: */
+#ifdef UNDEF
+static OSSL_FUNC_keymgmt_import_fn mlkem_import;
+static OSSL_FUNC_keymgmt_export_fn mlkem_export;
+static OSSL_FUNC_keymgmt_import_types_fn mlkem_imexport_types;
+static OSSL_FUNC_keymgmt_export_types_fn mlkem_imexport_types;
+static OSSL_FUNC_keymgmt_dup_fn mlkem_dup;
+#endif /* UNDEF */
+
+struct mlkem_gen_ctx {
+    void *provctx;
+    int selection;
+};
+
+static void *mlkem_new(void *provctx)
+{
+    MLKEM768_KEY *key = NULL;
+
+    debug_print("MLKEMKM new key req\n");
+    if (!ossl_prov_is_running())
+        return 0;
+
+    key = OPENSSL_zalloc(sizeof(MLKEM768_KEY));
+    if (key != NULL) {
+        key->keytype = MLKEM_KEY_TYPE_768; /* TODO(ML-KEM) any type */
+        key->provctx = provctx;
+        /*
+         * ideally, this is a one-time allocation and ctx that should be within the
+         * provider context: OK to move it there to improve performance?? It would be
+         * the first algorithmspecific context stored: Feels weird (TODO(ML-KEM)).
```

Comment:
Moving it there has implications. It brings back some problems which were present previously but are now no longer present. E.g. if providers get loaded/unloaded after the default provider then the cache might reference the wrong instances. Also if the global property query setting gets changed.

It might be a good idea to assess the performance impact of this, so we can understand whether it is worth it or not.

Reply:
Captured in https://github.com/openssl/openssl/issues/25879, so consider this dealt with for now.